### ISSUE TITLE
JUnit support trick (not a bug but doc advice)

### ISSUE THREAD

**Issue by agentgt** [reactive-streams/reactive-streams-jvm#479]
This is not a bug but advice on how to get better JUnit support. I wasn't sure where to put this but I figure even if closed it might be worthwhile to someone.

TestNG support in Eclipse and even to some extent IntelliJ isn't that good not to mention we already use JUnit so I have had great success with this.

Simple extend one of the verification tests and add a JUnit `TestRule` to handle TestNGs `SkipException` as well as override every test method and annotate with `@org.junit.Test`.

```java
public abstract class JUnitPublisherVerification<T> extends PublisherVerification<T> {
    public JUnitPublisherVerification(TestEnvironment env) {
        super(env);
    }
    public JUnitPublisherVerification(TestEnvironment env, long publisherReferenceGCTimeoutMillis) {
        super(env, publisherReferenceGCTimeoutMillis);
    }
    
    @Rule
    public TestRule skipRule = new TestRule() {
        public Statement apply(
                final Statement base,
                Description desc) {

            return new Statement() {
                public void evaluate() throws Throwable {
                    try {
                        base.evaluate();
                    }
                    catch (SkipException ex) {
                        Assume.assumeNoException(ex);
                    }
                }
            };
        }
    };
    
    @Override @Before
    public void setUp() throws Exception {
        super.setUp();
    }
    @Override @Test
    public void required_createPublisher1MustProduceAStreamOfExactly1Element() throws Throwable {
        super.required_createPublisher1MustProduceAStreamOfExactly1Element();
    }
   // snipped the other test methods
}
```

I tried some tricks of making JUnit use the TestNG annotations but while they worked (I think ... can't recall) for the build the IDE plugins would get confused. So that is the only tedious part.

I suppose I could make a separate project if folks are interested and have some magic unit test reflection on the verification interfaces to make sure the override methods are up to date.

**Comment by akarnokd** [comment]
It's higly unlikey the TCK would change the underlying test framework at this point and otherwise I don't have any trouble with having both JUnit and TestNG code in my projects using on the TCK.

**Comment by agentgt** [comment]
This isn't about compatibility issues with having JUnit and TestNG mixed. This is so you can use the JUnit Runner to run tests. The JUnit runner is better supported on most IDEs. For example TestNG has trouble running a single test on Eclipse.

The advice is not the for the library maintainers but for others who want to run JUnit directly on a TCK class.

Besides you still need TestNG anyway as the classes have TestNG annotations. Again I don't have a problem with this.

**Comment by akarnokd** [comment]
Not sure what Eclipse problem you are referring to. I have run individual classes and even test methods of TestNG for years with no problems. You install the latest [plugin](https://marketplace.eclipse.org/content/testng-eclipse) and make sure the target project has the latest stable TestNG library. So the question is, is there a particular problem or issue with running reactive-streams-tck to post such workarounds here?

![image](https://user-images.githubusercontent.com/1269832/74064053-1adfa880-49f2-11ea-99ca-f9e59dad94fa.png)

**Comment by agentgt** [comment]
I have forgotten which version of TestNG eclipse plugin but I was using the Eclipse version 2019-06 (can't remember the name).

When I did get it to work on an older Eclipse (can't remember which one) if I clicked on a single test and ran it it would not work correctly. 

Furthermore if I run my entire projects test suite it obviously doesn't aggregate the JUnit ones or TestNG ones together

Finally while combining TestNG with JUnit works in Gradle IIRC I had major issues getting it to work properly with Maven particularly multi-module maven projects. I could find all the issues but I don't care to as the latter problems of having two different test plugins and test results and corresponding UIs was annoying and not worth my time. 

Again I'm not complaining that the TCK is in TestNG... I'm just showing a workaround for folks that want to run it entirely in JUnit.

Anyway I did this work awhile ago with JUnit and thought perhaps someone else would find it useful if the had the same problems/concerns as I did.
